controllers/receiver_controller.go

```diff
@@ -18,172 +18,213 @@ package controllers
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// Record suspension metrics
-	defer r.RecordSuspend(ctx, receiver, receiver.Spec.Suspend)
-	// Return early if the object is suspended
-	if receiver.Spec.Suspend {
-		log.Info("Reconciliation is suspended for this object")
-		return ctrl.Result{}, nil
-	}
+	// Initialize the runtime patcher with the current version of the object.
+	patcher := patch.NewSerialPatcher(obj, r.Client)
 
-	// Initialize the patch helper
-	patchHelper, err := patch.NewHelper(receiver, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
 	defer func() {
-		// Patch the object, ignoring conflicts on the conditions owned by this controller
-		patchOpts := []patch.Option{
-			patch.WithOwnedConditions{
-				Conditions: []string{
-					meta.ReadyCondition,
-					meta.ReconcilingCondition,
-					meta.StalledCondition,
-				},
-			},
+		// Patch finalizers, status and conditions.
+		if err := r.patch(ctx, obj, patcher); err != nil {
+			retErr = kerrors.NewAggregate([]error{retErr, err})
 		}
 
-		// Determine if the resource is still being reconciled, or if it has stalled, and record this observation
-		if retErr == nil && (result.IsZero() || !result.Requeue) {
-			// We are no longer reconciling
-			conditions.Delete(receiver, meta.ReconcilingCondition)
-
-			// We have now observed this generation
-			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
-
-			readyCondition := conditions.Get(receiver, meta.ReadyCondition)
-			switch readyCondition.Status {
-			case metav1.ConditionFalse:
-				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
-				conditions.MarkStalled(receiver, readyCondition.Reason, readyCondition.Message)
-			case metav1.ConditionTrue:
-				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
-				conditions.Delete(receiver, meta.StalledCondition)
-			}
-		}
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
 
-		// Finally, patch the resource
-		if err := patchHelper.Patch(ctx, receiver, patchOpts...); err != nil {
-			retErr = errors.NewAggregate([]error{retErr, err})
+		// Emit warning event if the reconciliation failed.
+		if retErr != nil {
+			r.Event(obj, corev1.EventTypeWarning, meta.FailedReason, retErr.Error())
 		}
 
-		// Always record readiness and duration metrics
-		r.Metrics.RecordReadiness(ctx, receiver)
-		r.Metrics.RecordDuration(ctx, receiver, start)
-
+		// Log and emit success event.
+		if retErr == nil && conditions.IsReady(obj) {
+			msg := fmt.Sprintf("Reconciliation finished in %s, next run in %s",
+				time.Since(reconcileStart).String(), obj.Spec.Interval.Duration.String())
+			log.Info(msg)
+			r.Event(obj, corev1.EventTypeNormal, meta.SucceededReason, msg)
```

Comment:
Just another observation related to the earlier discussion about events. The above results in unique events like:
```
  Normal  Succeeded  22m    notification-controller  Reconciliation finished in 6.523649ms, next run in 10m0s
  Normal  Succeeded  12m    notification-controller  Reconciliation finished in 360.501µs, next run in 10m0s
  Normal  Succeeded  2m32s  notification-controller  Reconciliation finished in 279.767µs, next run in 10m0s
```
They are actually the same event and could be grouped together with `x times` next to it. But because the message contains execution time, each of these are unique events.
In SC, we no longer have such events or logs that show the execution time. Reconciliation of the same object without any change results in the same event with an increment to the event count.
Maybe it'd be better to do the same here. Keep the next run time but not the variable execution duration.

Reply:
Oh yeah, especially here where the reconciliation does no external calls, printing the execution time makes no sense.